research-prototype/Programs/Source/ipae2e.mpc

```diff
@@ -0,0 +1,196 @@
+    stepsize = 2 ** exponent
+    local_params['stopbit'] = stopbit
+    local_params['credit'] = credit
+    local_params['stepsize'] = stepsize
+
+    for current in rows[:-stepsize]:
+        successor = current + stepsize
+        # another optimization could be to multiply one time all the
+        # helperbit * istrigger bits since these are fixed.
+        if exponent == 0:
+            # credit is initialized to value because stopbit[successor] is always 1
+            # and reports[current][Columns.VALUE] = value[current] * stopbit[successor]
+            cmd = (
+                f"flag = helperbit_and_istrigger_flags[{successor}];"
+                f"credit[{current}] = "
+                f"reports[{current}][Columns.VALUE] +"
+                f"flag * reports[{successor}][Columns.VALUE];"
+                f"stopbit[{current}] = flag"
+            )
+        else:
+            cmd = (
+                f"flag = stopbits[{exponent-1}][{current}] "
+                f"* helperbit_and_istrigger_flags[{successor}];"
+                f"credit[{current}] = "
+                f"credit_arrays[{exponent-1}][{current}] + "
+                f"flag * credit_arrays[{exponent-1}][{successor}];"
+                f"stopbit[{current}] = flag * stopbits[{exponent-1}][{successor}]"
+            )
+        exec(cmd, {}, local_params)
+
+    if verbose:
+        print_ln('iteration: %s', exponent)
+        print_ln('credit %s', credit.reveal())
+        print_ln('stopbit %s', stopbit.reveal())
+
+# I can't sum all of credit, I need to sum only those that correspond to last touches
+# helperbit_xor_istrigger * credit_arrays[-1] will give this.
+final_credits = Array(numrows, sint)
+local_params['final_credits'] = final_credits
+
+for row in rows:
+    cmd = (
+        f"final_credits[{row}] = "
+        f"helperbit_xor_istrigger_flags[{row}] * credit_arrays[-1][{row}]"
+    )
+    exec(cmd, {}, local_params)
+
+if verbose:
+    for row in rows:
+        cmd = (
+            f"print_ln('%s, %s, %s, %s, %s', "
+            f"reports[{row}][Columns.MATCHKEY].reveal(), "
+            f"reports[{row}][Columns.IS_TRIGGER].reveal(),"
+            f"reports[{row}][Columns.VALUE].reveal(), "
+            f"reports[{row}][Columns.BREAKDOWN_KEY].reveal(),)"
+            f"final_credits[{row}].reveal())"
+        )
+        exec(cmd, {}, local_params)
+
+# SUM BREAKDOWNS
+# for now we have hard coded 4 breakdown keys
+breakdown_keys = list(range(4))
+
+breakdown_key_sums = Array(4, sint)
+local_params['breakdown_key_sums'] = breakdown_key_sums
+
+for breakdown_key in breakdown_keys:
+    cmd = f"breakdown_key_sums[{breakdown_key}] = sint(0)"
+    exec(cmd, {}, local_params)
+    for row in rows:
+        cmd = (
+            f"breakdown_key_sums[{breakdown_key}] = breakdown_key_sums[{breakdown_key}] + "
+            f"(reports[{row}][3] == {breakdown_key}).if_else(final_credits[{row}], 0)"
+        )
+        exec(cmd, {}, local_params)
```

Comment:
@bmcase I'm 99% sure this does the same thing as before (but doesn't do the "tree" approach. Given that we have to do the if_else on every row anyways, I'm not sure it really added any efficiency, and this is definitely easier to follow. Let me know if that makes sense.

Reply:
breakdown_key_sums[{breakdown_key}] = breakdown_key_sums[{breakdown_key}] + (reports[{row}][3] == {breakdown_key}).if_else(final_credits[{row}], 0)

- @eriktaubeneck  this looks like it should be the same; as long as this doesn't increase the rounds for you it should be fine.  It might be that the circuit you get at though won't run multi-threaded, but maybe we can test that out later.   
-can we use beakdown_key instead of "3" here?

Reply:
good catch!